File: Sismique/ClientApp/app/components/shared/categorytype.ts
import { Category, ICategoryProperties } from "./category";

export interface ICategoryTypeProperties {
    id: number;
    name: string;
    order: number;
    categories: ICategoryProperties[];
}

export class CategoryType {
    id: number;
    name: string;
    order: number;
    categories: Category[];

    deserialize(input: ICategoryTypeProperties): CategoryType {
        this.id = input.id;
        this.name = input.name;
        this.order = input.order;

        this.categories = [];
        if (input.categories != null) {
            for (let c of input.categories) {
                const category = new Category();
                category.deserialize(c);
                this.categories.push(category);
            }
        }

        return this;
    }
}

File: Sismique/Models/CategoryType.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace ARIO.Models
{
    public class CategoryType
    {
        public int ID { get; set; }

        [Required] public string Name { get; set; }

        [Required] public int Order { get; set; }

        public ICollection<Category> Categories { get; set; } = new List<Category>();
    }
}